### PR TITLE
NEW!! Providers (Indonesian)

### PR DESCRIPTION
**Pull request by hexated**
**1. Provider :** 
  - Otakudesu (https://otakudesu.tube)
  - AnimeIndo (https://anime-indo.one)
  - Multiplex (https://146.19.24.137)
  - Idlix (https://193.178.172.113)

**2. Add Extractors**
**3. Small fix**

### CODE REVIEW

**Comment by hexated** [comment]
I am trying to extract link from gdrive.to but I stuck with [this](https://justpaste.it/7ub4j)..
do you know how to decrypt that..?

(link => https://gdriveplayer.to/embed2.php?link=8qZOnmTPDgv6zJOCBKPR4QPNNhZF8ByQBtxRO24%252BQVh48tCbCIofINZFT7sqSGdoz83WDUCjPsfTxJBAfg%252FQxsfF0KBIpEf0KiTdC4OBdSe7JkaKX%252BZPrt13k%252BTUeSsyCUB6BYFd5fRPVpVz5wyxTNrNTzQ6qSaia21lQx1HV7uqEgKRVg9mNY0L13SzTWgvzc6%252BYQpML19z27BuSPyO3%252F)

**Comment by LagradOst** [comment]
Looks like AES encryption with iv and salt

File: app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimeIndoProvider.kt [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        val title = this.selectFirst("div.title")!!.text().trim()
+        val href = getProperAnimeLink(this.selectFirst("a")!!.attr("href"))
+        val posterUrl = this.select("img[itemprop=image]").attr("src").toString()
+        val type = getType(this.select("div.type").text().trim())
+        val epNum =
+            this.selectFirst("span.episode")?.ownText()?.replace(Regex("[^0-9]"), "")?.trim()
+                ?.toIntOrNull()
+        return newAnimeSearchResponse(title, href, type) {
+            this.posterUrl = posterUrl
+            addDubStatus(dubExist = false, subExist = true, subEpisodes = epNum)
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimeIndoProvider.kt:83]: I would prefer you use addSub here instead

File: app/src/main/java/com/lagradost/cloudstream3/animeproviders/OtakudesuProvider.kt [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+    private fun Element.toSearchResult(): SearchResponse {
+        val title = this.selectFirst("h2.jdlflm")!!.text().trim()
+        val href = this.selectFirst("a")!!.attr("href")
+        val posterUrl = this.select("div.thumbz > img").attr("src").toString()
+        val epNum = this.selectFirst("div.epz")?.ownText()?.replace(Regex("[^0-9]"), "")?.trim()
+            ?.toIntOrNull()
+        return newAnimeSearchResponse(title, href, TvType.Anime) {
+            this.posterUrl = posterUrl
+            addDubStatus(dubExist = false, subExist = true, subEpisodes = epNum)
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/cloudstream3/animeproviders/OtakudesuProvider.kt:67]: addSub

File: app/src/main/java/com/lagradost/cloudstream3/movieproviders/IdlixProvider.kt [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+
+        val year = Regex(",\\s?(\\d+)").find(
+            document.select("span.date").text().trim()
+        )?.groupValues?.get(1).toString().toIntOrNull()
+        val tvType = if (document.select("ul#section > li:nth-child(1)").text().contains("Episodes")
+        ) TvType.TvSeries else TvType.Movie
+        val description = document.select("div.wp-content > p").text().trim()
+        val rating =
+            document.selectFirst("span.dt_rating_vgs")?.text()?.toRatingInt()
+        val actors = document.select("div.persons > div[itemprop=actor]").map {
```
[the remark this answers is not in the document]
Comment (LagradOst) [reply app/src/main/java/com/lagradost/cloudstream3/movieproviders/IdlixProvider.kt:100]: not needed to change, just wanted to point out addActors can take a lot of different inputs

File: app/src/main/java/com/lagradost/cloudstream3/movieproviders/MultiplexProvider.kt [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    private fun Element.toSearchResult(): SearchResponse {
+        val title = this.selectFirst("h2.entry-title > a")!!.text().trim()
+        val href = this.selectFirst("a")!!.attr("href")
+        val posterUrl = fixUrl(this.selectFirst("a > img")?.attr("data-src").toString())
+        val quality = getQualityFromString(this.select("div.gmr-quality-item > a").text().trim())
+        return if (quality == null) {
+            val episode = this.select("div.gmr-numbeps > span").text().toIntOrNull()
+            newAnimeSearchResponse(title, href, TvType.TvSeries) {
+                this.posterUrl = posterUrl
+                addDubStatus(dubExist = false, subExist = true, subEpisodes = episode)
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/cloudstream3/movieproviders/MultiplexProvider.kt:66]: addSub

**Review by LagradOst — APPROVED** [review]
minor changes, no change needed

**Comment by Blatzar** [comment]
> I am trying to extract link from gdrive.to but I stuck with [this](https://justpaste.it/7ub4j).. do you know how to decrypt that..?

With debugger it's easy to find this code snippet which is what you wanna replicate

```js
var host = window.location.hostname;
if (host == 'gdriveplayer.co' || host == 'gdriveplayer.us' || host == 'gdriveplayer.org' || host == 'gdriveplayer.me' || host == 'gdriveplayer.io' || host == 'gdriveplayer.to' || host == 'gdriveplayer.fun' || host == 'gdriveplayer.app' || host == 'gdriveplayerapi.com' || host.includes('zeydhan.me')) {
} else {
  jwplayer().remove();
}
var pass = 'alsfheafsjklNIWORNiolNIOWNKLNXakjsfwnBdwjbwfkjbJjkopfjweopjASoiwnrflakefneiofrt';
var CryptoJSAesJson = {
  stringify: function (cipherParams) {
    var j = {
      ct: cipherParams.ciphertext.toString(CryptoJS.enc.Base64)
    };
    if (cipherParams.iv) j.iv = cipherParams.iv.toString();
    if (cipherParams.salt) j.s = cipherParams.salt.toString();
    return JSON.stringify(j);
  },
  parse: function (jsonStr) {
    var j = JSON.parse(jsonStr);
    var cipherParams = CryptoJS.lib.CipherParams.create({
      ciphertext: CryptoJS.enc.Base64.parse(j.ct)
    });
    if (j.iv) cipherParams.iv = CryptoJS.enc.Hex.parse(j.iv)
    if (j.s) cipherParams.salt = CryptoJS.enc.Hex.parse(j.s)
    return cipherParams;
  }
}
eval(JSON.parse(CryptoJS.AES.decrypt(data, pass, {
  format: CryptoJSAesJson
}).toString(CryptoJS.enc.Utf8)));
```

image

**Comment by hexated** [comment]
Feel free to merge

**Comment by LagradOst** [comment]
This pr also includes IndexSubtitleApi from https://github.com/LagradOst/CloudStream-3/pull/1193 do you want to finish indexsubtitle and merge both at the same time in the same pr? or do you want to remove all the files not related to the provider and merge?

Also did you solve the encryption?

**Comment by hexated** [comment]
> This pr also includes IndexSubtitleApi from #1193 do you want to finish indexsubtitle and merge both at the same time in the same pr? or do you want to remove all the files not related to the provider and merge?
> 
> Also did you solve the encryption?

indexsubtitle is done now. i just haven't implemented it yet, since i think the best experience for user is they can choose what subtitle provider they want rather load all in the same time, and i don't want to bother with the ui.

about cryptoJS. at first i thought only need salt and iv to decrypt text, but it also need password
172691922-df092ed6-bcb0-44b5-bf05-811228a7b20b
i don't know how to get that

**Comment by hexated** [comment]
question : is CS3 support to extract url like [this..](https://so-trym.topphimmoi.org/hlspm/0deb5c0d407df6d8fa92e863a080f37d)?

```
{
  "2048p": {
    "allowCache": true,
    "discontinuityStarts": [],
    "segments": [
      {
        "du": "5.333333",
        "link": "https://privatenpm.xyz/7cd510e29ca7b0f2c4d83dd139c664da.html"
      },
      {
        "du": "5.366667",
        "link": "https://privatenpm.xyz/3ea2e6fa040fd912a9379832331beb3c.html"
      },
      {
        "du": "6",
        "link": "https://privatenpm.xyz/91bb8f574cd5b9431c8f5fad75236fd2.html"
      },
      {
        "du": "3.7",
        "link": "https://privatenpm.xyz/9ae8edf39c87e30550377633f01372ff.html"
      }
    ],
    "targetDuration": 7,
    "mediaSequence": 0,
    "discontinuitySequence": 0,
    "endList": true,
    "md5": "0deb5c0d407df6d8fa92e863a080f37d"
  }
}
```

**Comment by Blatzar** [comment]
It should be plenty possible, the file ending does not matter.

**Comment by Blatzar** [comment]
Unable to get linked media working with ConcatenatingMediaSource :confused: 
Might write a custom parser, but that's going to take some time :thinking: 

How mergable is this? Is anything missing/incomplete?

**Comment by hexated** [comment]
No, this pr is already done..
That link for another provider i havent push

File: app/src/main/java/com/lagradost/cloudstream3/movieproviders/RebahinProvider.kt [hunk 5]
```diff
@@ -106,6 +107,7 @@ class RebahinProvider : MainAPI() {
         )?.groupValues?.get(1).toString().toIntOrNull()
         val tvType = if (url.contains("/series/")) TvType.TvSeries else TvType.Movie
         val description = document.select("span[itemprop=reviewBody] > p").text().trim()
+        val trailer = fixUrl(document.selectFirst("div.modal-body-trailer iframe")!!.attr("src"))
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/cloudstream3/movieproviders/RebahinProvider.kt:110]: fixUrlNull   ?.attr

File: app/src/main/java/com/lagradost/cloudstream3/syncproviders/providers/IndexSubtitleApi.kt [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+import com.lagradost.cloudstream3.*
+import com.lagradost.cloudstream3.subtitles.AbstractSubProvider
+import com.lagradost.cloudstream3.subtitles.AbstractSubtitleEntities
+import com.lagradost.cloudstream3.syncproviders.InAppAuthAPIManager
+import com.lagradost.cloudstream3.utils.SubtitleHelper
+
+class IndexSubtitleApi(index: Int) : InAppAuthAPIManager(index), AbstractSubProvider {
+    override val idPrefix = "indexsubtitle"
+    override val createAccountUrl = ""
+    override val requiresLogin = false
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/cloudstream3/syncproviders/providers/IndexSubtitleApi.kt:13]: if this has no login function impelemnted then dont use InAppAuthAPIManager, but AuthAPI instead

**Review by LagradOst — COMMENTED** [review]
👍

**Review by LagradOst — APPROVED** [review]
LGTM, ready to merge?

**Comment by hexated** [comment]
> LGTM, ready to merge?

yes..

**Pull request merged**